### PR TITLE
Bday

### PR DESCRIPTION
**Pull request by DrSharky**
Made birthday addition.

### CODE REVIEW

**Comment by pikulet** [comment]
Closes #446

File: routes/shop.js [hunk 1]
```diff
@@ -36,6 +36,16 @@ const shopItems = [
 
         }
     },
+    {
+        name: "Birthday Change",
+        desc: "Change your birthday once per purchase",
+        key: "bdayChange",
+        price: 10,
```
Comment (pikulet) [inline routes/shop.js:43]: I personally feel that birthday does not need to be a purchase but this works too xD

File: routes/user.js [hunk 2]
```diff
@@ -577,6 +578,44 @@ router.post("/avatar", async function (req, res) {
     }
 });
 
+router.post("/birthday", async function (req, res){
+    res.setHeader("Content-Type", "application/json");
+    try{
+        let userId = await routeUtils.verifyLoggedIn(req);
+        let prop = String(req.body.prop);
+        var itemsOwned = await redis.getUserItemsOwned(userId);
+        var perm = "changeBday";
```
Comment (pikulet) [inline routes/user.js:587]: maybe skip the permission check?

I reloaded backend + frontend, still got this permission error

how did you add permission for your user?
the concern is that all existing users won't have the bday permission

image

File: routes/user.js [hunk 3]
```diff
@@ -577,6 +578,44 @@ router.post("/avatar", async function (req, res) {
...
+        }
+
+        await models.User.updateOne(
+            { id: userId },
+            {
+                $set: { birthday: value, bdayChanged: true },
+                $inc: { "itemsOwned.bdayChange": -1 }
+            }
+        ).exec();
+        await redis.cacheUserInfo(userId, true);
```
Comment (pikulet) [inline routes/user.js:607]: 2023-03-20T13:47:17.255Z error: Cast to date failed for value "undefined" (type string) at path "birthday"
```
CastError: Cast to date failed for value "undefined" (type string) at path "birthday"
    at model.Query.exec (C:\Users\hui_j\beyondmaf\node_modules\mongoose\lib\query.js:4498:21)     
    at C:\Users\hui_j\beyondmaf\routes\user.js:606:11
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

File: react_main/src/components/Form.jsx [hunk 4]
```diff
@@ -189,6 +201,56 @@ export default function Form(props) {
...
+				else {
+					currentValue = new Date(field.value);
+				}
+				return (
+					<div className={fieldWrapperClass} key={field.ref}>
+						<div className="label">
+							{field.label}
+						</div>
+						<DatePicker
+							calendarAriaLabel="Toggle calendar"
```
[the remark this answers is not in the document]
Comment (DrSharky) [reply react_main/src/components/Form.jsx:223]: - For future consideration -

File: routes/shop.js (hunk 1 above)
Comment (DrSharky) [reply routes/shop.js:43]: - For future consideration -

File: routes/user.js (hunk 3 above)
Comment (DrSharky) [reply routes/user.js:607]: - Future fix

File: react_main/src/Main.jsx [hunk 5]
```diff
@@ -95,6 +95,14 @@ function Main() {
                     ), "basic", true);
                 }
 
+                if (res.data.bdayChanged == false) {
```
Comment (plokky) [inline react_main/src/Main.jsx:98]: This makes it pop up every single time you load the website if you don't set your birthday?

Comment (DrSharky) [reply react_main/src/Main.jsx:98]: Yeah, I guess that could be taken out. True.

**Review by pikulet — APPROVED** [review]
left comments that can be considered for future
- do not let user configure the year
- display month as string instead of number to avoid mmdd confusion
- make birthday changeable only once per account

**Review by plokky — APPROVED** [review]

**Review by pikulet — APPROVED** [review]

**Comment by pikulet** [comment]
happy birthday @DrSharky

**Pull request merged**